abn/application/applicationmap.go

```diff
@@ -0,0 +1,176 @@
+package application
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/iter8-tools/iter8/base/log"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	SECRET_POSTFIX string = ".iter8abnmetrics"
+)
+
+// ThreadSafeApplicationMap is type to control thread safety of operations on an application map
+type ThreadSafeApplicationMap struct {
+	mutex sync.Mutex
+	apps  map[string]*Application
+	rw    *ApplicationReaderWriter
+}
+
+// Applications is map of app name to Application
+// This is a global variable used to maintain an internal representation of the applications in a cluster
+var Applications = ThreadSafeApplicationMap{
+	apps: map[string]*Application{},
+}
+
+// Lock locks the application map; should always be followed by an Unlock()
+func (m *ThreadSafeApplicationMap) Lock() {
+	m.mutex.Lock()
+}
+
+// Unlock unlocks the application map
+func (m *ThreadSafeApplicationMap) Unlock() {
+	m.mutex.Unlock()
+}
+
+// Clear the application map
+func (m *ThreadSafeApplicationMap) Clear() {
```

Comment:
Left Add() as a map method; refactored all adds of new maps to use this method
Moved Clear() to test_helper since used only in tests
Refactored locks to be application specific as well as map specific. Map locks now in Get/Add/Clear; no longer assume caller calls them.
This seemed simpler.